### PR TITLE
Payloader

### PR DESCRIPTION
**Pull request by andpodob**

### CODE REVIEW

File: lib/rtp_vp9/payload_descriptor.ex [hunk 1]
```diff
@@ -123,6 +129,76 @@ defmodule Membrane.RTP.VP9.PayloadDescriptor do
...
+    first_octet <> picture_id <> tid_u_sid_d <> p_diffs <> tl0picidx <> scalability_structure
+  end
+
+  def serialize(%ScalabilityStructure{} = ss) do
+    first_octet = ss.first_octet
+
+    dimensions = Enum.reduce(ss.dimensions, <<>>, &(&2 <> serialize(&1)))
+
+    {n_g, pg_descriptions} =
+      if ss.pg_descriptions do
```
Comment (mat-hek) [inline lib/rtp_vp9/payload_descriptor.ex]: this should always be list

File: lib/rtp_vp9/payload_descriptor.ex [hunk 2]
```diff
@@ -123,6 +129,76 @@ defmodule Membrane.RTP.VP9.PayloadDescriptor do
...
+    tid_u_sid_d =
+      if l == 1,
+        do:
+          <<payload_descriptor.tid::3, payload_descriptor.u::1, payload_descriptor.sid::3,
+            payload_descriptor.d::1>>,
+        else: <<>>
+
+    p_diffs =
+      if p == 1 and f == 1,
+        do: payload_descriptor.p_diffs |> Enum.reduce(<<>>, &(&2 <> <<&1>>)),
```
Comment (mat-hek) [inline lib/rtp_vp9/payload_descriptor.ex]: use Enum.map_join

File: lib/rtp_vp9/payload_descriptor.ex [hunk 3]
```diff
@@ -123,6 +129,76 @@ defmodule Membrane.RTP.VP9.PayloadDescriptor do
...
+    first_octet <> dimensions <> n_g <> pg_descriptions
+  end
+
+  def serialize(%SSDimension{} = dimension) do
+    <<dimension.width::16, dimension.height::16>>
+  end
+
+  def serialize(%PGDescription{} = pg_description) do
+    {r, p_diffs} =
+      if pg_description.p_diffs do
```
Comment (mat-hek) [inline lib/rtp_vp9/payload_descriptor.ex]: this should always be list

File: lib/rtp_vp9/payloader.ex [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+        state
+      ) do
+    chunk_count = ceil(byte_size(payload) / state.max_payload_size)
+    max_chunk_size = ceil(byte_size(payload) / chunk_count)
+
+    {buffers, _i} =
+      payload
+      |> Bunch.Binary.chunk_every_rem(max_chunk_size)
+      |> add_descriptors()
+      |> Enum.map_reduce(1, fn chunk, i ->
```
Comment (mat-hek) [inline lib/rtp_vp9/payloader.ex]: using `|> List.update_at(-1, ...)` would make it more clear

File: lib/rtp_vp9/payloader.ex [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+    [first_chunk | chunks] = chunks
+
+    chunks =
+      chunks
+      |> Enum.reduce([], fn chunk, acc ->
+        [@middle_fragment_descriptor <> chunk | acc]
+      end)
+
+    chunks = [@last_fragment_descriptor <> last_chunk | chunks]
+    [@first_fragment_descriptor <> first_chunk | Enum.reverse(chunks)]
```
Comment (mat-hek) [inline lib/rtp_vp9/payloader.ex]: these two clauses do pretty much the same thing, let's merge them

File: lib/rtp_vp9/payloader.ex [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+
+  @impl true
+  def handle_demand(:output, size, :buffers, _ctx, state) do
+    {{:ok, demand: {:input, size}}, state}
+  end
+
+  @impl true
+  def handle_process(
+        :input,
+        %Buffer{metadata: metadata, payload: payload},
```
Comment (mat-hek) [inline lib/rtp_vp9/payloader.ex]: let's do this match in the function body

**Review by mat-hek — CHANGES_REQUESTED** [review]

File: .gitignore [hunk 7]
```diff
@@ -1,3 +1,4 @@
+payloader_with_session_bin_test.exs
```
Comment (mat-hek) [inline .gitignore]: to remove

File: lib/rtp_vp9/payload_descriptor.ex [hunk 8]
```diff
@@ -170,31 +170,22 @@ defmodule Membrane.RTP.VP9.PayloadDescriptor do
   def serialize(%ScalabilityStructure{} = ss) do
     first_octet = ss.first_octet
 
-    dimensions = Enum.reduce(ss.dimensions, <<>>, &(&2 <> serialize(&1)))
+    dimensions = ss.dimensions |> Enum.map_join("", &serialize(&1))
```
Comment (mat-hek) [inline lib/rtp_vp9/payload_descriptor.ex]: the empty string is the default joiner

File: lib/rtp_vp9/payload_descriptor.ex [hunk 9]
```diff
@@ -170,31 +170,22 @@ defmodule Membrane.RTP.VP9.PayloadDescriptor do
...
 
-    {n_g, pg_descriptions} =
-      if ss.pg_descriptions do
-        {<<length(ss.pg_descriptions)>>,
-         Enum.reduce(ss.pg_descriptions, <<>>, &(&2 <> serialize(&1)))}
-      else
-        {<<0>>, <<>>}
-      end
+    n_g = length(ss.pg_descriptions)
```
Comment (mat-hek) [inline lib/rtp_vp9/payload_descriptor.ex]:
```suggestion
    n_g = <<length(ss.pg_descriptions)>>
```

File: lib/rtp_vp9/payloader.ex [hunk 10]
```diff
@@ -69,37 +69,39 @@ defmodule Membrane.RTP.VP9.Payloader do
   @impl true
   def handle_process(
         :input,
-        %Buffer{metadata: metadata, payload: payload},
+        buffer,
```
Comment (mat-hek) [inline lib/rtp_vp9/payloader.ex]: all arguments will fit in a single line now

File: lib/rtp_vp9/payloader.ex [hunk 11]
```diff
@@ -69,37 +69,39 @@ defmodule Membrane.RTP.VP9.Payloader do
...
         _ctx,
         state
       ) do
+    %Buffer{metadata: metadata, payload: payload} = buffer
     chunk_count = ceil(byte_size(payload) / state.max_payload_size)
     max_chunk_size = ceil(byte_size(payload) / chunk_count)
```
Comment (mat-hek) [inline lib/rtp_vp9/payloader.ex:73]: doesn't this equal max_payload_size?

Comment (andpodob) [reply lib/rtp_vp9/payloader.ex:73]: We discussed that in VP8 payloader.

**Review by mat-hek — APPROVED** [review]

**Pull request merged**